cwltool/workflow_job.py

```diff
@@ -586,6 +607,113 @@ def receive_output(
         else:
             _logger.info("[%s] completed %s", step.name, processStatus)
 
+        if "http://commonwl.org/cwltool#Loop" in requirements:
```

Comment:
There's a lot going on here.

Architecturally, I'd like to see this go into its own class, the way the scatter methods use ReceiveScatterOutput.  I'll call in `LoopStep` for the sake of discussion.

This would let you encapsulate the loop logic into a class instead of adding a large amount of new code to a method that was previously much shorter.

I would also like to see the loop logic go into a function that returns a JobsGenerator, similar to the way scatter works.

So the sketch of the way this would work is:

1. in `try_make_job`, if there's a loop, it calls `jobs = LoopStep().job(step, inputobj, callback, runtimeContext)`  This creates a LoopStep object and calls the job generator method on it.
2. in `LoopStep.job()` it checks the loop condition.
3. LoopStep has a method called loop_callback
4. if the condition passes, it yields the first iteration `step.job(inputobj, self.loop_callback, runtimeContext)`
5. When that finishes, loop_callback is called with the result of the step.  Record the results, update the input object, and set a flag that an iteration has completed
6. Control will periodically come back to `LoopStep.job` which will check that flag.  If there's nothing ready yet, return None
7. If an iteration completed, we go back to the top and check the loop condition.  If the loop condition is still true, yield step.job() with the updated inputobj
8. If the loop condition is false, call the _original_ callback with the final results